File: stack/856.py
class Solution:
    def scoreOfParentheses(self, S):
        """
        :type S: str
        :rtype: int
        """
        
        stack = [0]
        for c in S:
            if c == '(':
                stack.append(0)
            else:
                temp = stack.pop()
                if temp == 0:
                    ret = 1
                else:
                    ret = temp * 2
                stack[-1] += ret
        return stack[0]

File: stack/856.cc
class Solution {
public:
    int scoreOfParentheses(string S) {
        std::stack<int> s;
        s.push(0);
        for (const auto &c : S) {
            if (c == '(') {
                s.push(0);
            } else {
                int cur = s.top();
                s.pop();
                int pre = s.top();
                s.pop();
                s.push(pre + std::max(2*cur, 1));
            }
        }
        return s.top();
    }
};
